pallets/api/src/extension.rs

```diff
@@ -136,17 +138,187 @@ fn version(env: &impl Environment) -> u8 {
+	}
+
+	#[test]
+	fn dispatch_call_log_target_works() {
+		assert!(matches!(
+			<DispatchCallLogTarget as LogTarget>::LOG_TARGET,
+			"pop-api::extension::dispatch"
+		));
+	}
+
+	#[test]
+	fn read_state_log_target_works() {
+		assert!(matches!(
+			<ReadStateLogTarget as LogTarget>::LOG_TARGET,
+			"pop-api::extension::read-state"
+		));
+	}
+
+	mod versioned_error {
+		use super::{RetVal::Converging, *};
+
+		// Mock versioned error.
+		#[derive(Debug)]
+		pub enum VersionedError {
+			V0(DispatchError),
+			V1(DispatchError),
+		}
+
+		impl TryFrom<(DispatchError, u8)> for VersionedError {
+			type Error = DispatchError;
+
+			fn try_from(value: (DispatchError, u8)) -> Result<Self> {
+				let (error, version) = value;
+				match version {
+					0 => Ok(VersionedError::V0(error)),
+					1 => Ok(VersionedError::V1(error)),
+					_ => Err(Other("DecodingFailed")),
+				}
+			}
+		}
+
+		impl From<VersionedError> for u32 {
+			// Mock conversion based on error and version.
+			fn from(value: VersionedError) -> Self {
+				match value {
+					VersionedError::V0(error) => match error {
+						BadOrigin => 1,
+						_ => 100,
+					},
+					VersionedError::V1(error) => match error {
+						BadOrigin => 2,
+						_ => 200,
+					},
+				}
+			}
+		}
+
+		#[test]
+		fn versioned_error_converter_works() {
+			for (version, error, expected_result) in vec![
+				(0, BadOrigin, 1),
+				(0, CannotLookup, 100),
+				(1, BadOrigin, 2),
+				(1, CannotLookup, 200),
+			] {
+				let env =
+					MockEnvironment { func_id: u16::from_le_bytes([0, version]), ext_id: 0u16 };
+				// Because `Retval` does not implement the `Debug` trait the result has to be
+				// unwrapped.
+				let Ok(Converging(result)) =
+					VersionedErrorConverter::<VersionedError>::convert(error, &env)
+				else {
+					panic!("should not happen")
+				};
+				assert_eq!(result, expected_result);
+			}
+		}
+
+		#[test]
+		fn versioned_error_converter_fails() {
+			let env = MockEnvironment { func_id: u16::from_le_bytes([0, 2]), ext_id: 0u16 };
+			let result = VersionedErrorConverter::<VersionedError>::convert(BadOrigin, &env).err();
+			assert_eq!(result, Some(Other("DecodingFailed")));
+		}
+	}
+
+	mod versioned_result {
+		use VersionedRuntimeResult::{V0, V1};
+
+		use super::*;
+
+		// Mock versioned runtime result.
+		#[derive(Debug, PartialEq)]
+		pub enum VersionedRuntimeResult {
+			V0(u8),
+			V1(u8),
+		}
+
+		impl TryFrom<(u8, u8)> for VersionedRuntimeResult {
+			type Error = DispatchError;
+
+			// Mock conversion based on result and version.
+			fn try_from(value: (u8, u8)) -> Result<Self> {
+				let (result, version) = value;
+				match version {
+					0 if result <= 50 => Ok(V0(result)),
+					0 if result > 50 => Ok(V0(50)),
+					1 if result <= 100 => Ok(V1(result)),
+					1 if result > 100 => Ok(V1(100)),
+					_ => Err(Other("DecodingFailed")),
+				}
+			}
+		}
+
+		#[test]
+		fn versioned_result_converter_works() {
+			for (version, value, expected_result) in vec![
```

Comment:
As above, add comment that describes that v0 logic is to cap a result at 50 and v1 is to increase the cap to 100.